File: cdm/java/mil/tatrc/physiology/datamodel/actions/SEAdvanceTime.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.actions;

import com.kitware.physiology.cdm.Actions.AdvanceTimeData;

import mil.tatrc.physiology.datamodel.properties.SEScalarTime;

public class SEAdvanceTime extends SEAction 
{
  protected SEScalarTime time;
  
  public SEAdvanceTime() 
  {
    time = null;
  }
  
  public void copy(SEAdvanceTime other)
  {
    if(this==other)
      return;
    super.copy(other);
  }
  
  public void reset() 
  {
    super.reset();
    if (hasTime())
      time.invalidate();
  }
  
  public boolean isValid()
  {
    return hasTime();
  }

  public static void load(AdvanceTimeData src, SEAdvanceTime dst) 
  {
    SEAction.load(src.getAction(), dst);
    if(src.hasTime())
      SEScalarTime.load(src.getTime(),dst.getTime());
  }
  public static AdvanceTimeData unload(SEAdvanceTime src)
  {
    AdvanceTimeData.Builder dst = AdvanceTimeData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEAdvanceTime src, AdvanceTimeData.Builder dst)
  {
    SEAction.unload(src, dst.getActionBuilder());
    if (src.hasTime())
      dst.setTime(SEScalarTime.unload(src.time));
  }
  
  public boolean hasTime()
  {
    return time == null ? false : time.isValid();
  }
  public SEScalarTime getTime() 
  {
    if (time == null)
      time = new SEScalarTime();
    return time;
  }
  
  public String toString() 
  {
    return "Advance Time" 
        + "\n\tTime: " + getTime();
  }
}

File: cdm/cpp/scenario/SEAdvanceTime.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "scenario/SEAdvanceTime.h"
#include "properties/SEScalarTime.h"

SEAdvanceTime::SEAdvanceTime() : SEAction()
{
  m_Time=nullptr;
}

SEAdvanceTime::~SEAdvanceTime()
{
  Clear();
}

void SEAdvanceTime::Clear()
{
  SEAction::Clear();
  SAFE_DELETE(m_Time);
}

bool SEAdvanceTime::IsValid() const
{ 
  return HasTime() && m_Time->IsValid();
}

void SEAdvanceTime::Load(const cdm::AdvanceTimeData& src, SEAdvanceTime& dst)
{
  SEAdvanceTime::Serialize(src, dst);
}
void SEAdvanceTime::Serialize(const cdm::AdvanceTimeData& src, SEAdvanceTime& dst)
{
  dst.Clear();
  if (src.has_time())
    SEScalarTime::Load(src.time(), dst.GetTime());
}

cdm::AdvanceTimeData* SEAdvanceTime::Unload(const SEAdvanceTime& src)
{
  cdm::AdvanceTimeData* dst = new cdm::AdvanceTimeData();
  SEAdvanceTime::Serialize(src, *dst);
  return dst;
}
void SEAdvanceTime::Serialize(const SEAdvanceTime& src, cdm::AdvanceTimeData& dst)
{
  if (src.HasTime())
    dst.set_allocated_time(SEScalarTime::Unload(*src.m_Time));
}

void SEAdvanceTime::ToString(std::ostream &str) const
{  
  if(HasComment())
    str<<"\n\tComment : "<<m_Comment;
  str << "Advance Time : " << m_Time;
}


bool SEAdvanceTime::HasTime() const
{
  return m_Time==nullptr?false:m_Time->IsValid();
}
SEScalarTime& SEAdvanceTime::GetTime()
{
  if(m_Time==nullptr)
    m_Time=new SEScalarTime();
  return *m_Time;
}
double SEAdvanceTime::GetTime(const TimeUnit& unit) const
{
  if (m_Time == nullptr)
    return SEScalar::dNaN();
  return m_Time->GetValue(unit);
}